docker-freeipa/bootstrap/start_freeipa_app.sh

```diff
@@ -28,4 +28,4 @@ if [ "$SECURE_RANDOM" == "false" ]; then
   FREEIPA_JAVA_OPTS="$FREEIPA_JAVA_OPTS -Djava.security.egd=file:/dev/./urandom"
 fi
 
-eval "(java $FREEIPA_JAVA_OPTS -jar /freeipa.jar) & JAVAPID=\$!; trap \"kill \$JAVAPID; wait \$JAVAPID\" SIGINT SIGTERM; wait \$JAVAPID"
+eval "(java $FREEIPA_JAVA_OPTS -cp /freeipa.jar:/classpath org.springframework.boot.loader.JarLauncher) & JAVAPID=\$!; trap \"kill \$JAVAPID; wait \$JAVAPID\" SIGINT SIGTERM; wait \$JAVAPID"
```

Comment:
god, what the hell.. why do we need this classpath modification hack when we can do it in a good way and put into resources directory?

Reply:
Thanks for providing feedback. Let me know if you want to setup meeting for a conversation. I feel like this is likely to turn into a long conversation chain.

I really wanted to avoid changing this line of code, so I want to describe the reason why specifying `-cp` and the class name may be a necessary compromise (or necessary "evil") for license compliance. But I should make it clear that I am not a lawyer but I have read through the GPL license and understand it from a software engineering perspective. But I do not understand the GPL from a case law perspective or whatever the legal term is.

This change adds a file `getkeytab.py` as a plugin to FreeIPA. The `getkeytab.py` plugin's purpose in this architecture is to provide the ability to create and retrieve keytabs for service principals. The FreeIPA built-in tool `ipa-getkeytab` (which happens to actually be used by this plugin) cannot be run in our environment without making more significant changes. `ipa-getkeytab` uses the LDAP network protocol and the current tunnel proxy, which is under active development, in this architecture currently only supports web protocols. Although improvements could be made in the future to support more protocols, the option adding an other protocols to the tunnel right now was ruled out due to the schedule impact. This change unblocks multiple teams that need to retrieve keytabs today and I do not want to delay these teams longer than necessary.

FreeIPA is licensed under GPL 3 and it runs python plugins in the same program (not using fork/exec). According to https://www.gnu.org/licenses/gpl-faq.en.html#GPLPlugins, if the main program doesn't use fork/exec then the plugin must be licensed under the GPL.

Given that `getkeytab.py` is licensed as GPL 3, I wanted to ensure that we comply with terms of it. Originally I wanted to create a separate web service that was not a plugin. This could have clearly avoided the GPL altogether, but this option was not chosen again due to the schedule timeline. My next idea was to create an RPM to install it. This option was also not chosen due to the schedule timeline. So it was a compromise to use SaltStack as the installation mechanism. Given that the decision was made (at least for now) to use both GPL code and use Salt to install it, it leaves some complications for compliance. I originally thought about packing `getkeytab.py` in the existing resources directory. This had the effect of packaging `getkeytab.py` in the resources inside of `freeipa.jar`. I wanted to avoid a situation where the entire FreeIPA management might be interpreted as needing to be licensed under the GPL. I am not a lawyer but it concerns me to have a GPL licensed file in a JAR that is not GPL licensed. Even though in this case the GPL file would not actually be executed in the java process, but it is just a resource.

So one question is packaging it in this way a ["mere aggregation"](https://www.gnu.org/licenses/gpl-faq.en.html#MereAggregation) which can be broken into a few questions:
1. Is having `getkeytab.py` in the same git repo is "mere aggregation"? According to https://www.gnu.org/licenses/gpl-faq.en.html#MereAggregation, yes
2. Is having `getkeytab.py` in a separate directory on a docker image a "mere aggregation"? According to https://www.gnu.org/licenses/gpl-faq.en.html#MereAggregation, yes
3. Is having `getkeytab.py` in the resource of a JAR a "mere aggregation"? According to https://www.gnu.org/licenses/gpl-faq.en.html#MereAggregation, "If the modules are included in the same executable file, they are definitely combined in one program" so the answer is no because the JAR is an executable file. It also states "The only condition is that you cannot release the aggregate under a license that prohibits users from exercising rights that each program's individual license would grant them." So there would need to be a way for anyone we distribute the `freeipa.jar` file to have an ability to modify the `getkeytab.py`. I suppose they could treat the JAR file like a zip file and "unzip", modify, and "rezip" it. But I don't really feel that is the right way to handle it, for example what happens if the JAR is signed or something like that?

As a result of item 3 (above), I wanted to avoid having `getkeytab.py` in the JAR. The current code which packages the resources to send for bootstrapping the SalkStack files, already pulls files from the classpath. I wanted to avoid complicating the logic in `BootstrapService.java` so I just updated the classpath. I didn't want to add `/classpath` to the manifest of the `freeipa.jar` because I think that is messier. So in order to add the GPL python resource `getkeytab.py` I had to remove the `-jar` option and add `-cp` and specify the main class. Yes it is messy. I agree, but it is a compromise.

---

As to the it will be a pain to use dev mode, by dev mode I am assuming you mean:
```
./gradlew :freeipa:bootRun -PjvmArgs="-Dfreeipa.db.addr=localhost \
-Dserver.port=8090 \
-Dvault.addr=localhost \
-Dvault.root.token=<VAULT_ROOT_TOKEN> \
-Dspring.config.location=$(pwd)/freeipa/src/main/resources/application.yml,$(pwd)/freeipa/build/resources/main/application.properties"
```
 I made changes to the `build.gradle` file to ensure the README.md commands work.

Reply:
@lacikaaa I think this comment thread is also applicable to your review comments:

"is this python plugin a public stuff or why is there a license file?"
`getkeytab.py` is licensed GPL because it is a plugin to the FreeIPA API server which is a GPL 3 program.

"please do not add files this way. Do not modify the docker-freeipa at all. Add your salt files to freeipa modul src/resources/freeipa-salt directory and you can use the file.managed state as you did and we did"
I described my license concerns in regard to the GPLS and a "mere aggregation" above in this thread.

Reply:
we agreed on @jamisonbennett will check if placing it on the freeipa image is a good solution: https://github.com/hortonworks/cloudbreak-images/tree/master/saltstack/freeipa
so we would avoid this dirty hack and still meet the licence requirements

Reply:
I had a discussion with the legal team. Since we created the plugin and the plugin is combined with the FreeIPA program inside the customer environment, we can license it as Apache in our environment. Its only when it gets combined/executed in the customer's environment that the license would convert to GPL. At that point the customer could perform modifications in their environment since they have the full source to this plugin (its just a python).

This clarification should get rid of the complexity that we wanted to avoid. And it should allow the `getkeytab.py` to be placed in the freeipa.jar's resource directory (the same as other resources - that way we don't have to worry about another repo such as https://github.com/hortonworks/cloudbreak-images/tree/master/saltstack/freeipa).

I will make these changes.

Reply:
I updated the pull request to address the comments in this thread and the comments from @lacikaaa . Thanks.